qiskit/algorithms/optimizers/de.py

```diff
@@ -0,0 +1,134 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Differential Evolution optimizer"""
+
+from typing import Callable, Tuple, List, Dict, Union
+
+import numpy as np
+from scipy.optimize import differential_evolution
+from .optimizer import Optimizer, OptimizerSupportLevel
+
+
+class DifferentialEvolution(Optimizer):
+    """
+    Differential Evolution optimizer
+
+    The Differential Evolution Optimizer is a global and derivative-free
+    optimizer that seeks to find the optimal value of an objective function,
+    f(x) (where x represents the vector of possible parameters) through
+    stochastically searching the optimization space of possible solutions.
+    It belongs to the family of evolutionary optimization techniques
+    such as genetic and particle swarm algorithms. It is best suited
+    for combinatorial optimization problems and operates on real-parameter
+    and real-valued objective functions.
+
+    Specifically, the algorithm begins with an initial population
+    of candidate solutions (can be random or user-specified) and new candidates
+    are iteratively created through a combination of mutation, recombination
+    and stochastic selection. The main parameters that govern the optimization
+    include the evolutionary strategy, population size, mutation constant
+    (differential weight) and recombination factor (crossover probability).
+    The algorithm terminates if either the maximum number of iterations/generations
+    have been reached or the standard deviation of the generation of candidates
+    is below a user-specified threshold. Specificially, the total number of
+    function evaluations equals (maxiter + 1)* popsize * len(x), without polishing
+    (see below).
+
+    Uses scipy.optimize.differential_evolution
+    For further detail, please refer to
+    https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.differential_evolution.html
+    """
+
+    _OPTIONS = ['maxiter', 'strategy', 'popsize', 'mutation',
+                'recombination', 'init', 'tol', 'atol',
+                'polish', 'disp', 'workers']
+
+    def __init__(self,
+                 maxiter: int = 10,
+                 strategy: str = 'currenttobest1bin',

```

Comment:
Thanks @woodsp-ibm for the feedback. I am happy to make any changes to make the code cleaner and clearer for the user. I can update the docstring related to the args that you mentioned.

In regards to the default values, I was using a different `strategy` for testing with a different problem, and for the reasons you described I used smaller values for `maxiter` and `popsize`. I can use the same default values as in the scipy documentation for consistency and clarity, and use the smaller default values for the unit tests. 

At a high level, setting `polish=True` will improve the optimization by using the L-BFGS-B method for the 'best' population member. However, this means that the maximum number of function evaluations is not easily predicted (and will be greater than `(maxiter + 1) * popsize * len(x)`). I was thinking that the default value being False would make this clearer, and the user can tweak the hyperparameters as needed. Otherwise, I can keep the default value to be True and update the docstrings as necessary; while indicating the tradeoff between function evaluations and accuracy. Let me know what you think.

Reply:
If you believe that its better to have polish=False as the default, for typical usages as we would have in Qiskit, then lets leave it as is. The tradeoff can always be stated in the documentation whatever the default value - I note at present there is info in the class docstring itself that says `see below` which I imagine is referring the reader to look at the `polish` parameter for that extra information what happens in that case. Perhaps things there can be improved to make it clearer for the reader in any case?